src/python/pants/core/goals/test.py

```diff
@@ -204,6 +224,155 @@ class TestFieldSet(FieldSet, metaclass=ABCMeta):
     __test__ = False
 
 
+_T = TypeVar("_T")
+_FieldSetT = TypeVar("_FieldSetT", bound=TestFieldSet)
+
+
+@dataclass(frozen=True)
+class Partition(Generic[_FieldSetT]):
+    field_sets: tuple[_FieldSetT, ...]
+    description: str
+
+
+@runtime_ignore_subscripts
+class Partitions(FrozenDict[str, Partition[_FieldSetT]]):
+    """A mapping from <partition key> to <partition>.
+
+    When implementing a test-runner, one of your rules will return this type, taking in a
+    `PartitionRequest` specific to your runner.
+
+    The return likely will fit into one of:
+        - Returning no empty partitions, if your tool is being skipped.
+        - Returning a singleton partition per input, if your tool can only run tests
+            from within one input per process.
+        - Returning partitions containing a variable number of inputs, if your tool
+            supports processing inputs in batches.
+
+    NOTE: The partition may be divided further into multiple sub-partitions.
```

Comment:
Do you mean into "multiple batches"? And that that happens via `test.py`?

Reply:
Good call-out.

I've deleted the extra sub-batching logic from this PR because everything returns partitions of size 1 and we don't have a config value for "max batch size" (yet). I'm a little worried about how re-batching will play with the use of `partition.key` as the `extra_output_prefix` - if the plugin returns a partition of 100 with a certain key, and `test.py` decides to split that into 5 batches of 20, how can it (safely) decide what the new key values should be on the `SubPartition`s?

I am tempted to say that we have `test.py` work with the exact partitions it receives, and leave any max-batch-size enforcement to the individual plugins that choose to implement batching (i.e. have `[pytest].max_batch_size` instead of `[test].max_batch_size`). That way plugin authors can rely on keys/descriptions not changing between the partitioning and test-running rules. What do you think?

Reply:
> I am tempted to say that we have test.py work with the exact partitions it receives, and leave any max-batch-size enforcement to the individual plugins that choose to implement batching (i.e. have [pytest].max_batch_size instead of [test].max_batch_size). That way plugin authors can rely on keys/descriptions not changing between the partitioning and test-running rules. What do you think?

I'm in general not super convinced on the batching idea in general, tbh. It seems surprising to users and could cause issues like duplicate database names. Given that non-Pants runs everything in a single batch/partition, us emulating that if you ask for it is reasonable imo.